### PR TITLE
refactor: Bitcoin txn validation & signing

### PR DESCRIPTION
**Pull request by ujjwal-cyph**
- [x] allow only supported input type
- [x] allow only supported output type
- [x] exactly one change output
- [x] change address validation
- [x] uppper limit on number of inputs & outputs
- [x] sign transaction inputs
- [x] send transaction script type detection

### CODE REVIEW

File: apps/btc_family/btc_txn.c [hunk 1]
```diff
@@ -443,14 +479,80 @@ static bool get_user_verification() {
...
-  // TODO: Sign each input and send the signature
-  return false;
+  // populate hashes cache for segwit transaction types
+  btc_segwit_init_cache(btc_txn_context);
+  if (!derive_hdnode_from_path(hd_path, 3, SECP256K1_NAME, buffer, &node) ||
+      false == validate_change_address(&node)) {
+    btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                   ERROR_DATA_FLOW_INVALID_DATA);
+    memzero(buffer, sizeof(buffer));
+    return false;
```
Comment (amanCypherock) [inline apps/btc_family/btc_txn.c:523]: Should clear hdnode as well?

File: apps/btc_family/btc_txn_helpers.c [hunk 2]
```diff
@@ -101,6 +103,40 @@
...
+/**
+ *
+ * @param context
+ * @param input_index
+ * @param digest
+ * @return
+ */
+STATIC void calculate_p2wpkh_digest(const btc_txn_context_t *context,
+                                    uint8_t input_index,
+                                    uint8_t *digest);
```
Comment (amanCypherock) [inline apps/btc_family/btc_txn_helpers.c:138]: Please populate comments

**Review by amanCypherock — CHANGES_REQUESTED** [review]
Please update these minor changs

File: apps/btc_family/btc_helpers.c [hunk 3]
```diff
@@ -199,3 +199,31 @@ bool btc_derivation_path_guard(const uint32_t *path, uint32_t depth) {
...
+  double fee_in_btc = 1.0 * value_in_sat / (SATOSHI_PER_BTC);
+  snprintf(msg, msg_len, "%0.*f %s", precision, fee_in_btc, g_app->lunit_name);
+}
+
+void btc_sig_to_script_sig(const uint8_t *sig,
+                           const uint8_t *pub_key,
+                           uint8_t *script_sig) {
+  if (NULL == sig || NULL == pub_key || NULL == script_sig) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply apps/btc_family/btc_helpers.c]: added a return code

File: apps/btc_family/btc_txn_helpers.c [hunk 4]
```diff
@@ -157,13 +193,140 @@ STATIC uint32_t get_transaction_weight(const btc_txn_context_t *txn_ctx) {
...
+  // double hash
+  sha256_Final(&sha_256_ctx, digest);
+  sha256_Raw(digest, 32, digest);
+  memzero(&sha_256_ctx, sizeof(sha_256_ctx));
+}
+
+STATIC void calculate_p2wpkh_digest(const btc_txn_context_t *context,
+                                    const uint8_t input_index,
+                                    uint8_t *digest) {
+  if (!context->segwit_cache.filled) {
```
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply apps/btc_family/btc_txn_helpers.c:277]: returning with a boolean now. this is now translated to the host as `unknown_error`. did not use ASSERT.

File: apps/btc_family/btc_helpers.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply apps/btc_family/btc_helpers.c]: moved this function `btc_script.c`

File: apps/btc_family/btc_txn_helpers.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply apps/btc_family/btc_txn_helpers.c]: added

File: apps/btc_family/btc_txn.c (hunk 1 above)
Comment (ujjwal-cyph) [reply apps/btc_family/btc_txn.c]: added

**Review by kartikCypherock — APPROVED** [review]

**Review by amanCypherock — APPROVED** [review]

**Pull request merged**